### PR TITLE
Collect metadata in k8s cm

### PR DESCRIPTION
**Pull request by mickymiek**
This PR gives an options to persist RBD and CephFS metadata (currently stored on node) in a k8s configmap.

Related task: #66

(Copied from [#102](https://github.com/ceph/ceph-csi/pull/102))

### CODE REVIEW

File: deploy/cephfs/kubernetes/csi-cephfsplugin.yaml [hunk 1]
```diff
@@ -49,11 +49,16 @@ spec:
             - "--endpoint=$(CSI_ENDPOINT)"
             - "--v=5"
             - "--drivername=csi-cephfsplugin"
+            - "--metadatastorage=node"
           env:
             - name: NODE_ID
               valueFrom:
                 fieldRef:
                   fieldPath: spec.nodeName
+            - name: POD_NAMESPACE
```
Comment (gman0) [inline deploy/cephfs/kubernetes/csi-cephfsplugin.yaml]: Add a quick description of this in the docs

File: deploy/rbd/kubernetes/csi-rbdplugin.yaml [hunk 2]
```diff
@@ -50,14 +50,19 @@ spec:
...
+            - "--containerized=true"
+            - "--metadatastorage=node"
           env:
             - name: HOST_ROOTFS
               value: "/rootfs"
             - name: NODE_ID
               valueFrom:
                 fieldRef:
                   fieldPath: spec.nodeName
+            - name: POD_NAMESPACE
```
Comment (gman0) [inline deploy/rbd/kubernetes/csi-rbdplugin.yaml]: Add a quick description of `POD_NAMESPACE` env var in the docs

File: deploy/rbd/kubernetes/csi-rbdplugin.yaml [hunk 3]
```diff
@@ -50,14 +50,19 @@ spec:
             - "--endpoint=$(CSI_ENDPOINT)"
             - "--v=5"
             - "--drivername=csi-rbdplugin"
-            - "--containerized=true"            
+            - "--containerized=true"
+            - "--metadatastorage=node"
```
Comment (gman0) [inline deploy/rbd/kubernetes/csi-rbdplugin.yaml]: IMHO k8s deployments should default to `--metadatastorage=k8s_configmap` @rootfs ?

**Comment by gman0** [comment]
LGTM! Besides the default metadata storage for k8s deployments I mentioned earlier. @rootfs I think this is ready to be merged once the last comment is addressed - if you don't oppose.

File: deploy/rbd/kubernetes/csi-rbdplugin.yaml (hunk 3 above)
Comment (rootfs) [reply deploy/rbd/kubernetes/csi-rbdplugin.yaml]: yes, please make that change @mickymiek 
btw, sorry for late response.

Comment (mickymiek) [reply deploy/rbd/kubernetes/csi-rbdplugin.yaml]: @rootfs done

**Comment by gman0** [comment]
Cool, lgtm

**Comment by rootfs** [comment]
thanks @mickymiek for getting this done and @gman0 for the review

**Comment by zhucan** [comment]
> Cool, lgtm

how to config "KUBERNETES_CONFIG_PATH"?

**Comment by gman0** [comment]
@zhucan according to the deployment documentation, it's an (optional) environment variable and may be set in the DaemonSet manifest of the plugin.

**Comment by zhucan** [comment]
> @zhucan according to the deployment documentation, it's an (optional) environment variable and may be set in the DaemonSet manifest of the plugin.

yeah, I know, but I want to know the value of the "KUBERNETES_CONFIG_PATH"，can you tell me how to set it ?

**Comment by gman0** [comment]
@zhucan this is only needed if you're deploying the plugin as out-of-cluster. In that case the variable should be set to the path of your `admin.conf` .

**Comment by zhucan** [comment]
> @zhucan this is only needed if you're deploying the plugin as out-of-cluster. In that case the variable should be set to the path of your `admin.conf` .

no, you are wrong；The value of the KUBERNETES_CONFIG_PATH is "~/.kube/config"

**Comment by zhucan** [comment]
> @zhucan this is only needed if you're deploying the plugin as out-of-cluster. In that case the variable should be set to the path of your `admin.conf` .

KUBERNETES_CONFIG_PATH: if you use k8s_configmap as metadata store, specify the path of your k8s config file (if not specified, the plugin will assume you're running it inside a k8s cluster and find the config itself).

**Comment by zhucan** [comment]
There is no examples for the csi-rbdplugin.yaml to set the KUBERNETES_CONFIG_PATH?

**Comment by zhucan** [comment]
> @zhucan
> 
> > no, you are wrong；The value of the KUBERNETES_CONFIG_PATH is "~/.kube/config"
> 
> sure, if that works for you
> 
> > There is no examples for the csi-rbdplugin.yaml to set the KUBERNETES_CONFIG_PATH?
> 
> Unfortunately not ATM, you could try and add something like this in your RBD plugin deployment in the `csi-rbdplugin` container spec definition:
> 
> ```yaml
>         - name: csi-rbdplugin
>           ...
>           env:
>             - name: KUBERNETES_CONFIG_PATH
>               value: "/path-to-your-k8s-config"
>           ...
> ```

when the KUBERNETES_CONFIG_PATH not set;  There is an error:  Failed to get cluster config with error: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined

**Comment by zhucan** [comment]
> > @zhucan
> > > no, you are wrong；The value of the KUBERNETES_CONFIG_PATH is "~/.kube/config"
> > 
> > 
> > sure, if that works for you
> > > There is no examples for the csi-rbdplugin.yaml to set the KUBERNETES_CONFIG_PATH?
> > 
> > 
> > Unfortunately not ATM, you could try and add something like this in your RBD plugin deployment in the `csi-rbdplugin` container spec definition:
> > ```yaml
> >         - name: csi-rbdplugin
> >           ...
> >           env:
> >             - name: KUBERNETES_CONFIG_PATH
> >               value: "/path-to-your-k8s-config"
> >           ...
> > ```
> 
> when the KUBERNETES_CONFIG_PATH not set; There is an error: Failed to get cluster config with error: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined

I deploy the plugin in the k8s-cluster

**Comment by Madhu-1** [comment]
@zhucan AFAIK if you are deploying in k8s-cluster you don't need to set this env, just remove it from the container yaml file.

**Comment by zhucan** [comment]
> @zhucan AFAIK if you are deploying in k8s-cluster you don't need to set this env, just remove it from the container yaml file.

I don't set it，it will start from "rest.InClusterConfig()", There is an error
__________________________________________________
func NewK8sClient() *k8s.Clientset {
	var cfg *rest.Config
	var err error
	cPath := os.Getenv("KUBERNETES_CONFIG_PATH")
	if cPath != "" {
		cfg, err = clientcmd.BuildConfigFromFlags("", cPath)
		if err != nil {
			glog.Errorf("Failed to get cluster config with error: %v\n", err)
			os.Exit(1)
		}
	} else {
		cfg, err = rest.InClusterConfig()
		if err != nil {
			glog.Errorf("Failed to get cluster config with error: %v\n", err)
			os.Exit(1)
		}
	}
	client, err := k8s.NewForConfig(cfg)
	if err != nil {
		glog.Errorf("Failed to create client with error: %v\n", err)
		os.Exit(1)
	}
	return client
}

**Comment by zhucan** [comment]
> @zhucan could you please open up a new issue describing this problem + other info that might aid in debugging?

ok, wait for a moment, it maybe a new issue

**Comment by Madhu-1** [comment]
> > @zhucan AFAIK if you are deploying in k8s-cluster you don't need to set this env, just remove it from the container yaml file.
> 
> I don't set it，it will start from "rest.InClusterConfig()", There is an error
> 
> func NewK8sClient() *k8s.Clientset {
> var cfg *rest.Config
> var err error
> cPath := os.Getenv("KUBERNETES_CONFIG_PATH")
> if cPath != "" {
> cfg, err = clientcmd.BuildConfigFromFlags("", cPath)
> if err != nil {
> glog.Errorf("Failed to get cluster config with error: %v\n", err)
> os.Exit(1)
> }
> } else {
> cfg, err = rest.InClusterConfig()
> if err != nil {
> glog.Errorf("Failed to get cluster config with error: %v\n", err)
> os.Exit(1)
> }
> }
> client, err := k8s.NewForConfig(cfg)
> if err != nil {
> glog.Errorf("Failed to create client with error: %v\n", err)
> os.Exit(1)
> }
> return client
> }

do you have this two files in your cluster
 "/var/run/secrets/kubernetes.io/serviceaccount/token"
 "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"

**Comment by zhucan** [comment]
> > > @zhucan AFAIK if you are deploying in k8s-cluster you don't need to set this env, just remove it from the container yaml file.
> > 
> > 
> > I don't set it，it will start from "rest.InClusterConfig()", There is an error
> > func NewK8sClient() *k8s.Clientset {
> > var cfg *rest.Config
> > var err error
> > cPath := os.Getenv("KUBERNETES_CONFIG_PATH")
> > if cPath != "" {
> > cfg, err = clientcmd.BuildConfigFromFlags("", cPath)
> > if err != nil {
> > glog.Errorf("Failed to get cluster config with error: %v\n", err)
> > os.Exit(1)
> > }
> > } else {
> > cfg, err = rest.InClusterConfig()
> > if err != nil {
> > glog.Errorf("Failed to get cluster config with error: %v\n", err)
> > os.Exit(1)
> > }
> > }
> > client, err := k8s.NewForConfig(cfg)
> > if err != nil {
> > glog.Errorf("Failed to create client with error: %v\n", err)
> > os.Exit(1)
> > }
> > return client
> > }
> 
> do you have this two files in your cluster
> "/var/run/secrets/kubernetes.io/serviceaccount/token"
> "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"


no,no,no; the errors  in this part "	
host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
	if len(host) == 0 || len(port) == 0 {
		return nil, ErrNotInCluster
	}"

**Comment by Madhu-1** [comment]
https://github.com/coreos/etcd-operator/issues/731 
https://github.com/kubernetes/kubernetes/issues/40973

**Comment by zhucan** [comment]
> [coreos/etcd-operator#731](https://github.com/coreos/etcd-operator/issues/731)
> [kubernetes/kubernetes#40973](https://github.com/kubernetes/kubernetes/issues/40973)

yeah , How can I get the  "/var/run/secrets/kubernetes.io/serviceaccount/token" file ？

**Pull request merged**